### ISSUE TITLE
RCS1165 seems misleading (perhaps should not be enabled by default)

### ISSUE THREAD

**Issue by franklin-ross** [JosefPihrt/Roslynator#384]
Often when I write a generic function which should work on a value of any type, I will want to check whether reference type values are null and do something else, like throw an `ArgumentNullException`. Here's my most common use case:

```C#
public void DoSomething<T>(T value)
{
    if (value == null)
        throw new ArgumentNullException(nameof(value));

    // Do some things.
}
```

Which gives me the warning "Unconstrained type parameter checked for null".

As far as I know, this code is correct and will compile into optimal code in all cases. When used with a value type, (I believe) the compiler will remove that code, which is desirable. The suggested code fix, which changes it to `EqualityComparer<T>.Default.Equals(value, default(T))` is not semantically the same thing, and can't be elided by the compiler. I almost never want to check whether the value is default instead of null, because structs must behave reasonably in their default state. Furthermore using null perfectly describes my intent: if the value is null then throw, default has nothing to do with it.

When googling that error text, the top result is this [SO post](https://stackoverflow.com/questions/12396457/checking-instance-of-non-class-constrained-type-parameter-for-null-in-generic-me), which has a convoluted accepted answer, and which I think leads people further astray.

A common way people get around this error is to cast to object before checking for null, but as far as I'm concerned that's a terrible idea because you're now introducing a box for value types, which is the kind of performance hit generics help avoid.

Another way might be to create 2 different methods with different names, one with a constraint for reference types and another for value types. I don't believe this is a correct solution either if the code is otherwise identical, it complicates the API for no good reason.

For these reasons I believe this warning should be disabled by default. The code fix is fine to leave, even if it changes the semantics of the code which I find a little surprising. It seems to be warning me about something which at worst is not incorrect and, in at least one very common case, is actually the most correct option.

Thoughts? Am I missing some glaring issue this warning catches?

**Comment by franklin-ross** [comment]
Perhaps my assumption that the compiler will elide a comparison between a value type and null is wrong? I have always assumed that would be a simple optimisation, considering structs are always sealed and generics with struct parameters get their own implementation, so the compiler has all the information it needs.

When I look at the IL I can see a box instruction, which surprises me.

**Comment by franklin-ross** [comment]
So basically, I no longer know the right answer to this. If someone smarter than me has a definitive answer then perhaps the warning message for this could be enhanced to explain why it's bad.

**Comment by Dreamescaper** [comment]
Looking at [SharpLab](https://sharplab.io/#v2:C4LglgNgPgAgDAAhgRgNwFgBQMDMSBMCAwggN5YKUIAOATmAG4CGwApkgCwIAiA9gMoBbYAAsAPABUAfAAoJCZhACUFKuUxVNCMADMZihAF5DCAHYBXCMvVbbCUbV4B3M6xcA5SxABKrHa1pWUwBjVgBRAA9Q6mAwXlMZACIAIwgmRKUMDTsAX1VKPMwcoA=), I'd say that boxing always happens for generic types for null comparison. ( `box !!T`)

Still, I'd say that case that you've described is perfectly valid, and casting to object is completely redundant, so I'd like it to be disabled by default as well.

**Comment by LesRamer** [comment]
I think the test should be changed:

```csharp
public void DoSomething<T>(T value)
{
    if (!(value is T))
        throw new ArgumentNullException(nameof(value));

    // Do some things.
}
```
In this case "is" is special -- passing a null value in for a reference type will cause the ArgumentNullException to be thrown. There is no boxing that happens for value types, and I believe the JIT optimizes the test and all away for value-types.

**Comment by franklin-ross** [comment]
@LesRamer: That's a great workaround! It does seem to be equivalent to the original test, and avoids the box.

I wouldn't say it's super clear why someone would write this unless they'd read this thread, but in the absence of a better solution, I think this should be the code fix.

**Comment by franklin-ross** [comment]
Fair enough; but can I ask whether it's possible to have both fixes, as each seems right in different circumstances? There are times when I've run a code fix and been confused about the result, then gone and read some supporting documentation and learned some really interesting quirks of a language. I think this falls exactly into that category.

Going forward, I will be writing `!(value is T)` for null checks like this, as I don't believe there's any other way to be sure to avoid a box in all cases, especially across compilers.

**Comment by franklin-ross** [comment]
Realistically I'd just love the language to allow `x` to be a potentially non-nullable value type in `x is null`. That seems perfect, because it wouldn't allow any funkiness with equality overrides, and reads very clearly.

**Comment by Dreamescaper** [comment]
@wiz0u
Language specific suggestions should go here:
https://github.com/dotnet/csharplang
